### PR TITLE
Add inverse Box-Cox transformation

### LINKED ISSUE #119: RFC: compute the inverse Box-Cox transformation
## Checklist

> Please ensure the following tasks are completed before filing an issue.

-   [x] Read and understood the [Code of Conduct][code-of-conduct].
-   [x] Searched for existing issues and pull requests.
-   [x] If this is a general question, searched the [FAQ][faq] for an existing answer.
-   [x] If this is a feature request, the issue name begins with `RFC:`.

## Description

> Description of the issue (or feature request).

Add support for computing the inverse of the [Box-Cox transformation](https://en.wikipedia.org/wiki/Power_transform#Box.E2.80.93Cox_transformation).

Package: `@stdlib/math/base/special/boxcoxinv`
Alias: `base.boxcoxinv`

The implementation should be a straightforward branch which follows the canonical equation.

## Related Issues

> Does this issue (or feature request) have any related issues?

No.

## Questions

> Any questions for reviewers?

No.

## Other

> Any other information relevant to this issue (or feature request)? This may include screenshots, references, stack traces, sample output, and/or implementation notes.

Reference implementations:

-   SciPy \[[1](https://github.com/scipy/scipy/blob/2526df72e5d4ca8bad6e2f4b3cbdfbc33e805865/scipy/special/_boxcox.pxd#L30)\]\[[2](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.inv_boxcox.html)\]

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[faq]: https://github.com/stdlib-js/stdlib/blob/develop/FAQ.md

### PR DESCRIPTION
**Pull request by justin1dennison**
Resolves #119.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [x] Read and understand the [Code of Conduct][code-of-conduct].
-   [x] Read and understood the [licensing terms][license].
-   [x] Searched for existing issues and pull requests **before** submitting this pull request.
-   [x] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [x] Rebased onto latest `develop`.
-   [x] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request:

-   Implements the inverse of the one-parameter Box-Cox transformation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #119 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

### CODE REVIEW

**Comment by kgryte** [comment]
@justin1dennison Awesome! I will review this shortly!

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+# boxcoxinv
+
+> Compute the inverse of a one-parameter [Box-Cox transformation][box-cox-transformation].
+
+<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->
+
+<section class="intro">
+
+Compute the inverse of a one-parameter [Box-Cox transformation][box-cox-transformation].
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
To compute the inverse of a one-parameter [Box-Cox transformation][box-cox-transformation], one finds the `x` such that 
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+> Compute the inverse of a one-parameter [Box-Cox transformation][box-cox-transformation].
+
+<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->
+
+<section class="intro">
+
+Compute the inverse of a one-parameter [Box-Cox transformation][box-cox-transformation]. 
+
+<!-- <equation class="equation" label="eq:boxcox_transformation_one_parameter" align="center" raw="y^{\lambda} = \begin{cases}\frac{y^{\lambda} - 1}{\lambda} & \textrm{if}\ \lambda \neq 0 \\ \ln(y) & \textrm{if}\ \lambda = 0 \end{cases}" alt="One-Parameter Box-Cox Transformation"> -->
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
<!-- <equation class="equation" label="eq:inverse_boxcox_transformation_one_parameter" align="center" raw="y = \begin{cases}\frac{x^{\lambda} - 1}{\lambda} & \textrm{if}\ \lambda \neq 0 \\ \ln(x) & \textrm{if}\ \lambda = 0 \end{cases}" alt="Inverse One-Parameter Box-Cox Transformation"> -->
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+<div class="equation" align="center" data-raw-text="y^{\lambda} = \begin{cases}\frac{y^{\lambda} - 1}{\lambda} & \textrm{if}\ \lambda \neq 0 \\ \ln(y) & \textrm{if}\ \lambda = 0 \end{cases}" data-equation="eq:boxcox_transformation_one_parameter">
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
<div class="equation" align="center" data-raw-text="y = \begin{cases}\frac{x^{\lambda} - 1}{\lambda} & \textrm{if}\ \lambda \neq 0 \\ \ln(x) & \textrm{if}\ \lambda = 0 \end{cases}" data-equation="eq:inverse_boxcox_transformation_one_parameter">
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+    <img src="https://cdn.rawgit.com/stdlib-js/stdlib/1cd982bdbe87913bbb66e9d6a79b0acb0fb89616/lib/node_modules/@stdlib/math/base/special/boxcox/docs/img/equation_boxcox_transformation_one_parameter.svg" alt="One-Parameter Box-Cox Transformation" />
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
    <img src="" alt="Inverse One-Parameter Box-Cox Transformation" />
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    <br>
+</div>
+
+<!-- </equation> -->
+
+Specifically, for a given `y` and `{\lambda}` it finds the `x` such that `y = boxcox(x, {\lambda})`.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+
+<section class="usage">
+
+## Usage
+
+```javascript
+var boxcoxinv = require( '@stdlib/math/base/special/boxcoxinv' );
+```
+
+#### boxcoxinv( x, lambda )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
#### boxcoxinv( y, lambda )
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+## Examples
+
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var incrspace = require( '@stdlib/math/utils/incrspace' );
+var boxcoxinv = require( '@stdlib/math/base/special/boxcoxinv' );
+
+var x = incrspace( -1.0, 10.0, 1.0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
var y = incrspace( -1.0, 10.0, 1.0 );
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+var l = incrspace( -0.5, 5.0, 0.5 );
+var b;
+var i;
+var j;
+
+for ( i = 0; i < x.length; i++ ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
for ( i = 0; i < y.length; i++ ) {
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    for ( j = 0; j < l.length; j++ ) {
+        b = boxcoxinv( x[ i ], l[ j ] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
        b = boxcoxinv( y[ i ], l[ j ] );
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+        console.log( 'boxcoxinv(%d, %d) = %d', x[ i ], l[ j ], b );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
        console.log( 'boxcoxinv(%d, %d) = %d', y[ i ], l[ j ], b );
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/benchmark/python/benchmark.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Benchmark scipy.special.inv_boxcox."""
+
+from __future__ import print_function
+import timeit
+
+NAME = "inv_boxcox"
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/benchmark/python/benchmark.py]:
```suggestion
NAME = "boxcoxinv"
```

The typical convention is that the `NAME` variable should reflect the package name. This is so we can match benchmark TAP output with the corresponding JavaScript benchmark output.

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/docs/repl.txt [hunk 12]
```diff
@@ -0,0 +1,38 @@
+
+{{alias}}( x, lambda )
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/docs/repl.txt]:
```suggestion
{{alias}}( y, lambda )
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/docs/repl.txt [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    Computes the inverse of a one-parameter Box-Cox transformation.
+
+    Parameters
+    ----------
+    x: number
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/docs/repl.txt]:
```suggestion
    y: number
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/examples/index.js [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+'use strict';
+
+var incrspace = require( '@stdlib/math/utils/incrspace' );
+var boxcoxinv = require( './../lib' );
+
+var x = incrspace( -1.0, 10.0, 1.0 );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/examples/index.js]:
```suggestion
var y = incrspace( -1.0, 10.0, 1.0 );
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/examples/index.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+var l = incrspace( -0.5, 5.0, 0.5 );
+var b;
+var i;
+var j;
+
+for ( i = 0; i < x.length; i++ ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/examples/index.js]:
```suggestion
for ( i = 0; i < y.length; i++ ) {
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/examples/index.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+	for ( j = 0; j < l.length; j++ ) {
+		b = boxcoxinv( x[ i ], l[ j ] );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/examples/index.js]:
```suggestion
		b = boxcoxinv( y[ i ], l[ j ] );
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/examples/index.js [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+		console.log( 'boxcoxinv(%d, %d) = %d', x[ i ], l[ j ], b );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/examples/index.js]:
```suggestion
		console.log( 'boxcoxinv(%d, %d) = %d', y[ i ], l[ j ], b );
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+var isPositiveZero = require( '@stdlib/math/base/assert/is-positive-zero' );
+var isNegativeZero = require( '@stdlib/math/base/assert/is-negative-zero' );
+
+
+// MAIN //
+
+/**
+* Computes the inverse of a one-parameter Box-Cox transformation.
+*
+* @param {number} x - input value
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js]:
```suggestion
* @param {number} y - input value
```

The rationale here is for us to mirror the equation variable names.

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+* v = boxcoxinv( 0.0, -1.0 );
+* // returns 1.0
+*
+* v = boxcoxinv( 1.0, NaN );
+* // returns NaN
+*
+* v = boxcoxinv( NaN, 3.1 );
+* // returns NaN
+*/
+function boxcoxinv( x, lambda ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js]:
```suggestion
function boxcoxinv( y, lambda ) {
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+	if ( isnan( x ) || isnan( lambda ) ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js]:
```suggestion
	if ( isnan( y ) || isnan( lambda ) ) {
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+		return NaN;
+	}
+	if ( isPositiveZero( lambda ) || isNegativeZero( lambda ) ) {
+		return exp( x );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js]:
```suggestion
		return exp( y );
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js [hunk 22]
```diff
@@ -0,0 +1,77 @@
...
+	}
+	return exp( log1p( lambda * x ) / lambda );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js]:
```suggestion
	return exp( log1p( lambda*y ) / lambda );
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/package.json [hunk 23]
```diff
@@ -0,0 +1,62 @@
...
+    "stdmath",
+    "mathematics",
+    "math",
+    "power",
+    "transform",
+    "power transform",
+    "boxcox",
+    "transformation",
+    "box-cox"
+  ]
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/package.json:62]: Can add the keyword: `"inverse"`.

**Review by kgryte — CHANGES_REQUESTED** [review]
@justin1dennison Thanks for doing this! The only nitpicks are variable name changes in order to be consistent with the canonical equation.

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js [hunk 24]
```diff
@@ -0,0 +1,77 @@
...
+* // returns NaN
+*
+* v = boxcoxinv( NaN, 3.1 );
+* // returns NaN
+*/
+function boxcoxinv( x, lambda ) {
+	if ( isnan( x ) || isnan( lambda ) ) {
+		return NaN;
+	}
+	if ( isPositiveZero( lambda ) || isNegativeZero( lambda ) ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js]:
```suggestion
	if ( lambda === 0.0 ) {
```

A strict equality check suffices for both positive and negative zero.

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js [hunk 25]
```diff
@@ -0,0 +1,77 @@
...
+*/
+
+'use strict';
+
+// MODULES //
+
+var exp = require( '@stdlib/math/base/special/exp' );
+var log1p = require( '@stdlib/math/base/special/log1p' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var isPositiveZero = require( '@stdlib/math/base/assert/is-positive-zero' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js [hunk 26]
```diff
@@ -0,0 +1,77 @@
...
+var isNegativeZero = require( '@stdlib/math/base/assert/is-negative-zero' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/lib/main.js]:
```suggestion
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/benchmark/python/benchmark.py (hunk 11 above)
Comment (justin1dennison) [reply lib/node_modules/@stdlib/math/base/special/boxcoxinv/benchmark/python/benchmark.py]: That makes sense. This was the first one that I worked on that the `scipy` function differs in name.

Comment (kgryte) [reply lib/node_modules/@stdlib/math/base/special/boxcoxinv/benchmark/python/benchmark.py]: Yeah, without docs (our failing) or a big picture POV, it's impossible to know about some of these conventions. We need to do better in removing the mystery of various conventions and practices.

**Comment by justin1dennison** [comment]
I will try to catch some of those small goofs earlier. Thanks again for the review and guidance. Let me know if there are any other changes that I should make.

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md [hunk 27]
```diff
@@ -0,0 +1,138 @@
...
+
+# boxcoxinv
+
+> Compute the inverse of a one-parameter [Box-Cox transformation][box-cox-transformation].
+
+<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->
+
+<section class="intro">
+
+To compute the inverse of a one-parameter [Box-Cox transformation][box-cox-transformation], on finds the `x` such that
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]:
```suggestion
To compute the inverse of a one-parameter [Box-Cox transformation][box-cox-transformation], one finds the `x` such that 
```

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/test/test.js [hunk 28]
```diff
@@ -0,0 +1,147 @@
...
+	b = boxcoxinv( 1.0, NaN );
+	t.equal( isnan( b ), true, 'returns NaN' );
+
+	b = boxcoxinv( NaN, NaN );
+	t.equal( isnan( b ), true, 'returns NaN' );
+
+	t.end();
+});
+
+tape( 'the function returns `NaN` if `x` is negative and `lambda` is one', function test( t ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/math/base/special/boxcoxinv/test/test.js]: For the function to return `NaN`, does `lambda` have to be `1`? What if `lambda = 10`? Then, based on the implementation, I would assume that the function returns `NaN` for any combination of `y` and `lambda` which is less than negative one.

This test should be clarified.

**Review by kgryte — CHANGES_REQUESTED** [review]
@justin1dennison Only a typo and some test clarification are needed. Once these are resolved, this should be ready for merge!

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md (hunk 27 above)
Comment (justin1dennison) [reply lib/node_modules/@stdlib/math/base/special/boxcoxinv/README.md]: Oops. 🤦‍♂️

File: lib/node_modules/@stdlib/math/base/special/boxcoxinv/test/test.js (hunk 28 above)
Comment (justin1dennison) [reply lib/node_modules/@stdlib/math/base/special/boxcoxinv/test/test.js]: You are absolutely correct. I have, hopefully, addressed those concerns. Additionally, I added additionally testing for ranges of inputs. I recognize that I need to get better at testing so thank you for bringing that to my attention.

**Review by kgryte — APPROVED** [review]
@justin1dennison Awesome work! Thanks for working on this! Merging...

**Comment by justin1dennison** [comment]
@kgryte I don't know the policy on this. However, I found another typo in this package. Should I open an issue or something else? Thank you. I am having a good amount of fun working on these.

**Comment by kgryte** [comment]
@justin1dennison Feel free to open another PR (based on the most recent `develop`) which fixes the typo. Thanks!

And great! We're glad that you're having a good experience. Your work is certainly helping move the project forward, and we really appreciate all your work!

**Pull request merged**